### ISSUE TITLE
Feature request: official API for webworker support

### ISSUE THREAD

**Issue by tjenkinson** [webpack/webpack#2951]
**I'm submitting a feature request**

**Webpack version:**
2 (and maybe 1 if feasible)

It would be a nice feature for there to be an official API which would enable support for web workers and could be used in a plugin such as [webworkify-webpack](https://github.com/borisirota/webworkify-webpack).

At the moment it looks like webworkify-webpack has to rely on webpack internals which may change in the future.

There is an [official webpack loader](https://github.com/webpack/worker-loader) for web workers but this will result in duplicate code as the webworker code is compiled separately. The webworkify-webpack plugin is nicer because it passes the module code that the webworker entry point needs at runtime, as well as the webpack internals so that the worker can require the modules it needs. This means if the worker code uses moduleA, moduleB and moduleC, and the main application javascript also uses these modules they will only be in the build once. With the official plugin they will be in the build twice (afaik).

**Comment by TheLarkInn** [comment]
Also, have you looked at the webpack `target: 'webworker'` property that can be set on your config? I have a decent sample repo that shows the difference in structure between all of the targets. [You can take a look here](https://github.com/TheLarkInn/compare-webpack-target-bundles) however I would pull down the repo and make changes to see how they reflect between targets. 

I'm not sure if this completely satisfies what you are referring to, but maybe this leaves room to improve the webworker target property, and if needed add more core plugin/loader support.

**Comment by tjenkinson** [comment]
@TheLarkInn that's interesting I didn't realise that option existed.

I don't think it solves this though as the target would need to be "web". I think either the code that "webworkify-webpack" uses needs to become a documented api, or maybe a webpack loader for web workers could be built into webpack allowing it to use internals without concern.

I like the way "webworkify-webpack" works my concern is just that it's having to use webpack internals to function.

**Comment by sokra** [comment]
There are documentated APIs for the used stuff:

``` js
var __webpack_require__ = arguments[2];
var sources = __webpack_require__.m;
var ENTRY_MODULE = __webpack_require__.s || "0";
var webworkifyWebpackModuleId = arguments[0].id || arguments[0].i;
```

=>

``` js
// __webpack_require__ is the function
var sources = __webpack_modules__;
var ENTRY_MODULE = require.main.id;
var webworkifyWebpackModuleId = module.id;
```

And you can use `require.resolve` to get the module id of a module:

``` js
var work = require("webworkify-webpack");

work(require.resolve("./worker"));
```

`require.resolve` makes the module available without executing it and returns the module id.

**Comment by tjenkinson** [comment]
Thanks @sokra I found the [documentation](https://webpack.github.io/docs/api-in-modules.html) now.

Also I probably should have tagged the plugin developer @borisirota before. (I'm still trying to get my head around how it works)

It looks like there's still some code there though that depends on internals like [this](https://github.com/borisirota/webworkify-webpack/blob/master/index.js#L7-L27)?

Maybe this can be provided by webpack as well?

**Comment by borisirota** [comment]
Hi guys, I wasn't familiar will all the documented api and it seems much better that the current hacky workaround. I'll dive into it soon and respond with some details.

For now, like @tjenkinson mentioned, there is no way I'm aware of to access in runtime (not sure its the best solution though) to a minimal webpack's bootstrap function => required in order to execute the new bundle (was built in runtime) which consist of webpack's modules in the webworker.

**Comment by borisirota** [comment]
@tjenkinson @TheLarkInn @sokra 

Hi guys, truly sorry it took so long. Was very busy since last time I wrote here.

v2 of [webworkify-webpack](https://github.com/borisirota/webworkify-webpack) is based on webpack's api wherever possible.

The next goal is to make it based fully on webpack's api and for that your input can be very valuable.
I created the [future-work](https://github.com/borisirota/webworkify-webpack#future-work) section and you are more than welcome to point me to the right direction.

Thanks !
